File: tests/token.test.ts
import { afterEach, beforeEach, describe, expect, test } from "bun:test";
import { mkdtempSync, rmSync, statSync, readFileSync, existsSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { SignJWT, decodeJwt } from "jose";
import type { Config } from "../src/config.ts";
import { DpopKeyManager, canonicalHtu } from "../src/dpop.ts";
import { TokenStore, type TokenSet } from "../src/store.ts";
import { DpopTokenClient } from "../src/oauth/token.ts";
import type { Endpoints } from "../src/oauth/discovery.ts";
import { logger as baseLogger, type Logger } from "../src/logger.ts";

const TOKEN_ENDPOINT = "https://okta.example.com/oauth2/v1/token";
const endpoints: Endpoints = {
  issuer: "https://okta.example.com",
  authorizationEndpoint: "https://okta.example.com/oauth2/v1/authorize",
  tokenEndpoint: TOKEN_ENDPOINT,
};

let home: string;

function cfg(overrides: Partial<Config> = {}): Config {
  return {
    ADAPTER_BASE_URL: "https://adapter.example.com",
    OKTA_CLIENT_ID: "cid",
    AGENT_ID: "agent-1",
    OKTA_REDIRECT_PORT: 0,
    OKTA_SCOPES: "openid offline_access",
    DPOP_ALG: "ES256",
    DPOP_KEY_MODE: "persistent",
    BRIDGE_HOME: home,
    HTTP_TIMEOUT_MS: 30000,
    LOG_LEVEL: "error",
    ...overrides,
  };
}

/** Mint a fake (HS256) access-token JWT; the client decodes without verifying. */
const SECRET = new TextEncoder().encode("test-secret-test-secret-test-secret-xx");
async function mintToken(claims: Record<string, unknown>): Promise<string> {
  return await new SignJWT(claims)
    .setProtectedHeader({ alg: "HS256" })
    .setIssuedAt()
    .sign(SECRET);
}

interface RecordedCall {
  url: string;
  init: RequestInit;
}

/** Sequenced mock fetch: each handler maps to one call (last handler repeats). */
function mockFetch(handlers: Array<() => Response | Promise<Response>>): {
  fetch: typeof fetch;
  calls: RecordedCall[];
} {
  const calls: RecordedCall[] = [];
  const fn = (async (url: Parameters<typeof fetch>[0], init?: RequestInit) => {
    calls.push({ url: String(url), init: init ?? {} });
    const h = handlers[Math.min(calls.length - 1, handlers.length - 1)]!;
    return await h();
  }) as unknown as typeof fetch;
  return { fetch: fn, calls };
}

function jsonResponse(
  body: unknown,
  status = 200,
  headers: Record<string, string> = {},
): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "Content-Type": "application/json", ...headers },
  });
}

/** Decode the DPoP proof JWT a recorded call carried. */
function proofOf(call: RecordedCall): Record<string, unknown> {
  const headers = call.init.headers as Record<string, string>;
  return decodeJwt(headers.DPoP!) as Record<string, unknown>;
}

function bodyParams(call: RecordedCall): URLSearchParams {
  return new URLSearchParams(call.init.body as string);
}

/** A logger that records emitted event names per level. */
function recordingLogger(): { logger: Logger; events: Record<string, string[]> } {
  const events: Record<string, string[]> = { info: [], warn: [], error: [], debug: [] };
  const make = (): Logger => ({
    log: (level, event) => events[level]!.push(event),
    debug: (event) => events.debug!.push(event),
    info: (event) => events.info!.push(event),
    warn: (event) => events.warn!.push(event),
    error: (event) => events.error!.push(event),
    withCorrelation: () => make(),
  });
  return { logger: make(), events };
}

beforeEach(() => {
  home = mkdtempSync(join(tmpdir(), "okta-mcp-bridge-tok-"));
});
afterEach(() => {
  rmSync(home, { recursive: true, force: true });
});

async function setup(handlers: Array<() => Response | Promise<Response>>, logger?: Logger) {
  const km = await DpopKeyManager.create(cfg(), baseLogger);
  const store = new TokenStore(home);
  const { fetch, calls } = mockFetch(handlers);
  const client = new DpopTokenClient(cfg(), endpoints, km, store, logger ?? baseLogger, {
    fetch,
  });
  return { km, store, client, calls };
}

const CODE = { code: "the-code", redirectUri: "http://127.0.0.1:5555/callback", verifier: "v".repeat(43) };

describe("DpopTokenClient nonce handshake", () => {
  test("first /token use_dpop_nonce → retry succeeds in exactly two fetches, second proof has the nonce", async () => {
    const { km, client, calls } = await setup([
      () => jsonResponse({ error: "use_dpop_nonce" }, 400, { "DPoP-Nonce": "nonce-1" }),
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
          scope: "openid offline_access",
        }),
    ]);
    await client.exchangeCode(CODE);
    expect(calls.length).toBe(2);
    expect(proofOf(calls[0]!).nonce).toBeUndefined();
    expect(proofOf(calls[1]!).nonce).toBe("nonce-1");
  });

  test("token-request proof has htm POST and canonical htu == tokenEndpoint", async () => {
    const { km, client, calls } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    await client.exchangeCode(CODE);
    const proof = proofOf(calls[0]!);
    expect(proof.htm).toBe("POST");
    expect(proof.htu).toBe(canonicalHtu(TOKEN_ENDPOINT));
  });

  test("a later use_dpop_nonce triggers another single retry", async () => {
    const { km, client, store, calls } = await setup([
      // exchange: challenge then success
      () => jsonResponse({ error: "use_dpop_nonce" }, 400, { "DPoP-Nonce": "nonce-1" }),
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
          refresh_token: "rt-1",
        }),
      // refresh: Okta rotated the nonce → challenge again, then success
      () => jsonResponse({ error: "use_dpop_nonce" }, 400, { "DPoP-Nonce": "nonce-2" }),
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    const set = await client.exchangeCode(CODE);
    await client.refresh(set);
    expect(calls.length).toBe(4);
    expect(proofOf(calls[3]!).nonce).toBe("nonce-2");
    expect(store.load()).not.toBeNull();
  });
});

describe("DpopTokenClient exchange + binding", () => {
  test("exchangeCode persists a TokenSet with jkt == keyManager.jkt()", async () => {
    const { km, client, store } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    const set = await client.exchangeCode(CODE);
    expect(set.jkt).toBe(await km.jkt());
    const loaded = store.load();
    expect(loaded?.jkt).toBe(await km.jkt());
    expect(loaded?.accessToken).toBe(set.accessToken);
  });

  test("token_type != DPoP warns but does not throw", async () => {
    const { logger, events } = recordingLogger();
    const { km, client } = await setup(
      [
        async () =>
          jsonResponse({
            access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
            token_type: "Bearer",
            expires_in: 3600,
          }),
      ],
      logger,
    );
    await client.exchangeCode(CODE);
    expect(events.warn).toContain("oauth.token.not_dpop_bound");
  });

  test("cnf.jkt mismatch throws", async () => {
    const { client } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: "WRONG-THUMBPRINT" } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    await expect(client.exchangeCode(CODE)).rejects.toThrow(/cnf\.jkt/);
  });

  test("absent cnf does not throw", async () => {
    const { client } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ sub: "user" }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    const set = await client.exchangeCode(CODE);
    expect(set.accessToken).toBeDefined();
  });

  test("tokens.json is chmod 600 and not plaintext", async () => {
    const { km, client } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    const set = await client.exchangeCode(CODE);
    const path = join(home, "tokens.json");
    expect(statSync(path).mode & 0o777).toBe(0o600);
    const raw = readFileSync(path, "utf8");
    expect(raw).not.toContain(set.accessToken);
    const parsed = JSON.parse(raw) as Record<string, unknown>;
    expect(parsed.ciphertext).toBeDefined();
    expect(parsed.iv).toBeDefined();
    expect(parsed.tag).toBeDefined();
  });
});

describe("DpopTokenClient refresh + getAccessToken", () => {
  test("refresh sends grant_type=refresh_token with a DPoP proof and persists", async () => {
    const { km, client, store, calls } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    const prior: TokenSet = {
      accessToken: "old",
      refreshToken: "rt-1",
      tokenType: "DPoP",
      expiresAt: 0,
      scope: "openid offline_access",
      jkt: await km.jkt(),
    };
    const next = await client.refresh(prior);
    const params = bodyParams(calls[0]!);
    expect(params.get("grant_type")).toBe("refresh_token");
    expect(params.get("refresh_token")).toBe("rt-1");
    expect((calls[0]!.init.headers as Record<string, string>).DPoP).toBeDefined();
    // refresh_token carried forward when not rotated
    expect(next.refreshToken).toBe("rt-1");
    expect(store.load()?.accessToken).toBe(next.accessToken);
  });

  test("getAccessToken: valid stored token → no network", async () => {
    const { km, client, store, calls } = await setup([
      () => jsonResponse({ error: "should_not_be_called" }, 500),
    ]);
    store.save({
      accessToken: "still-good",
      tokenType: "DPoP",
      expiresAt: Math.floor(Date.now() / 1000) + 3600,
      scope: "openid",
      jkt: await km.jkt(),
    });
    const token = await client.getAccessToken(async () => {
      throw new Error("authorizeFn must not run");
    });
    expect(token).toBe("still-good");
    expect(calls.length).toBe(0);
  });

  test("getAccessToken: expired with refresh_token → refreshes", async () => {
    const { km, client, store, calls } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    store.save({
      accessToken: "expired",
      refreshToken: "rt-9",
      tokenType: "DPoP",
      expiresAt: Math.floor(Date.now() / 1000) - 10,
      scope: "openid",
      jkt: await km.jkt(),
    });
    const token = await client.getAccessToken(async () => {
      throw new Error("authorizeFn must not run on refresh");
    });
    expect(token).not.toBe("expired");
    expect(bodyParams(calls[0]!).get("grant_type")).toBe("refresh_token");
  });

  test("getAccessToken: no stored token → authorizeFn + exchange", async () => {
    const { km, client, calls } = await setup([
      async () =>
        jsonResponse({
          access_token: await mintToken({ cnf: { jkt: await km.jkt() } }),
          token_type: "DPoP",
          expires_in: 3600,
        }),
    ]);
    let authorized = false;
    const token = await client.getAccessToken(async () => {
      authorized = true;
      return CODE;
    });
    expect(authorized).toBe(true);
    expect(token).toBeDefined();
    expect(bodyParams(calls[0]!).get("grant_type")).toBe("authorization_code");
  });
});

File: internal/oauth/token_test.go
package oauth

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
	"testing"

	"github.com/joewitt99/bridge-mcp-client/internal/config"
	"github.com/joewitt99/bridge-mcp-client/internal/dpop"
	"github.com/joewitt99/bridge-mcp-client/internal/logx"
	"github.com/joewitt99/bridge-mcp-client/internal/store"
)

var codeResult = AuthCodeResult{Code: "the-code", RedirectURI: "http://127.0.0.1:0/cb", Verifier: strings.Repeat("v", 43)}

type recReq struct {
	url  string
	dpop string
	body url.Values
}

func seqDoer(t *testing.T, resps ...*http.Response) (Doer, *[]recReq) {
	t.Helper()
	calls := &[]recReq{}
	i := 0
	d := doerFunc(func(r *http.Request) (*http.Response, error) {
		b, _ := io.ReadAll(r.Body)
		v, _ := url.ParseQuery(string(b))
		*calls = append(*calls, recReq{url: r.URL.String(), dpop: r.Header.Get("DPoP"), body: v})
		if i >= len(resps) {
			t.Fatalf("unexpected token request #%d", i+1)
		}
		resp := resps[i]
		i++
		return resp, nil
	})
	return d, calls
}

func oauthResp(status int, body map[string]any, headers map[string]string) *http.Response {
	bb, _ := json.Marshal(body)
	h := http.Header{}
	for k, v := range headers {
		h.Set(k, v)
	}
	return &http.Response{StatusCode: status, Body: io.NopCloser(bytes.NewReader(bb)), Header: h}
}

func mintToken(claims map[string]any) string {
	pb, _ := json.Marshal(claims)
	return "e30." + base64.RawURLEncoding.EncodeToString(pb) + ".sig"
}

func successResp(km *dpop.KeyManager, extra map[string]any) *http.Response {
	body := map[string]any{
		"access_token": mintToken(map[string]any{"cnf": map[string]any{"jkt": km.JKT()}}),
		"token_type":   "DPoP",
		"expires_in":   3600,
		"scope":        "openid offline_access",
	}
	for k, v := range extra {
		body[k] = v
	}
	return oauthResp(200, body, nil)
}

func setup(t *testing.T, override string) (config.Config, *dpop.KeyManager, *store.TokenStore) {
	t.Helper()
	home := t.TempDir()
	env := map[string]string{
		"ADAPTER_BASE_URL": "https://adapter.example.com",
		"OKTA_CLIENT_ID":   "cid",
		"AGENT_ID":         "agent-1",
		"BRIDGE_HOME":      home,
		"LOG_LEVEL":        "error",
	}
	if override != "" {
		env["OKTA_TOKEN_DPOP_HTU"] = override
	}
	cfg, err := config.Load(env)
	if err != nil {
		t.Fatal(err)
	}
	km, err := dpop.NewKeyManager(cfg, nil)
	if err != nil {
		t.Fatal(err)
	}
	return cfg, km, store.New(home)
}

func proofPayload(t *testing.T, proof string) map[string]any {
	t.Helper()
	parts := strings.Split(proof, ".")
	if len(parts) != 3 {
		t.Fatalf("expected JWT proof, got %q", proof)
	}
	b, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		t.Fatal(err)
	}
	var m map[string]any
	if err := json.Unmarshal(b, &m); err != nil {
		t.Fatal(err)
	}
	return m
}

func TestNonceHandshake(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t,
		oauthResp(400, map[string]any{"error": "use_dpop_nonce"}, map[string]string{"DPoP-Nonce": "nonce-1"}),
		successResp(km, nil),
	)
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err != nil {
		t.Fatal(err)
	}
	if len(*calls) != 2 {
		t.Fatalf("expected exactly 2 token requests, got %d", len(*calls))
	}
	if _, ok := proofPayload(t, (*calls)[0].dpop)["nonce"]; ok {
		t.Error("first proof should have no nonce")
	}
	if proofPayload(t, (*calls)[1].dpop)["nonce"] != "nonce-1" {
		t.Error("retry proof should carry the nonce")
	}
}

func TestProofHtmHtu(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t, successResp(km, nil))
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err != nil {
		t.Fatal(err)
	}
	pl := proofPayload(t, (*calls)[0].dpop)
	wantHTU, _ := dpop.CanonicalHTU(ep.TokenEndpoint)
	if pl["htm"] != "POST" || pl["htu"] != wantHTU {
		t.Fatalf("proof htm/htu = %v / %v, want POST / %s", pl["htm"], pl["htu"], wantHTU)
	}
}

func TestExchangePersistsJKT(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, _ := seqDoer(t, successResp(km, nil))
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	set, err := c.ExchangeCode(context.Background(), codeResult)
	if err != nil {
		t.Fatal(err)
	}
	if set.JKT != km.JKT() {
		t.Fatalf("set.JKT = %s, want %s", set.JKT, km.JKT())
	}
	loaded, _ := st.Load()
	if loaded == nil || loaded.AccessToken != set.AccessToken {
		t.Fatal("token not persisted")
	}
}

func TestTokenTypeNotDPoPWarns(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, _ := seqDoer(t, successResp(km, map[string]any{"token_type": "Bearer"}))
	buf := &bytes.Buffer{}
	c := NewTokenClient(cfg, ep, km, st, logx.NewWith(buf, "warn"), doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(buf.String(), "oauth.token.not_dpop_bound") {
		t.Fatalf("expected not_dpop_bound warning; got %q", buf.String())
	}
}

func TestCnfMismatchThrows(t *testing.T) {
	cfg, km, st := setup(t, "")
	bad := oauthResp(200, map[string]any{
		"access_token": mintToken(map[string]any{"cnf": map[string]any{"jkt": "WRONG"}}),
		"token_type":   "DPoP",
		"expires_in":   3600,
	}, nil)
	doer, _ := seqDoer(t, bad)
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err == nil || !strings.Contains(err.Error(), "cnf.jkt") {
		t.Fatalf("expected cnf.jkt error, got %v", err)
	}
}

func TestCnfAbsentOK(t *testing.T) {
	cfg, km, st := setup(t, "")
	opaque := oauthResp(200, map[string]any{
		"access_token": mintToken(map[string]any{"sub": "user"}),
		"token_type":   "DPoP",
		"expires_in":   3600,
	}, nil)
	doer, _ := seqDoer(t, opaque)
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err != nil {
		t.Fatalf("absent cnf should not error: %v", err)
	}
}

func TestRefresh(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t, successResp(km, nil))
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	prior := store.TokenSet{AccessToken: "old", RefreshToken: "rt-1", TokenType: "DPoP", ExpiresAt: 0, Scope: "openid offline_access", JKT: km.JKT()}
	next, err := c.Refresh(context.Background(), prior)
	if err != nil {
		t.Fatal(err)
	}
	b := (*calls)[0].body
	if b.Get("grant_type") != "refresh_token" || b.Get("refresh_token") != "rt-1" {
		t.Fatalf("bad refresh params: %v", b)
	}
	if (*calls)[0].dpop == "" {
		t.Fatal("refresh must carry a DPoP proof")
	}
	if next.RefreshToken != "rt-1" {
		t.Fatalf("refresh token not carried forward: %q", next.RefreshToken)
	}
}

func TestGetAccessTokenValidNoNetwork(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, _ := seqDoer(t) // any request → t.Fatal
	st.Save(store.TokenSet{AccessToken: "still-good", TokenType: "DPoP", ExpiresAt: 9_999_999_999, Scope: "openid", JKT: km.JKT()})
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	tok, err := c.GetAccessToken(context.Background(), func() (AuthCodeResult, error) {
		t.Fatal("authFn must not run")
		return AuthCodeResult{}, nil
	})
	if err != nil || tok != "still-good" {
		t.Fatalf("got %q, %v", tok, err)
	}
}

func TestGetAccessTokenRefreshes(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t, successResp(km, nil))
	st.Save(store.TokenSet{AccessToken: "expired", RefreshToken: "rt-9", TokenType: "DPoP", ExpiresAt: 1, Scope: "openid", JKT: km.JKT()})
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	tok, err := c.GetAccessToken(context.Background(), func() (AuthCodeResult, error) {
		t.Fatal("authFn must not run on refresh")
		return AuthCodeResult{}, nil
	})
	if err != nil || tok == "expired" {
		t.Fatalf("expected refreshed token, got %q (%v)", tok, err)
	}
	if (*calls)[0].body.Get("grant_type") != "refresh_token" {
		t.Fatalf("expected refresh, got %v", (*calls)[0].body)
	}
}

func TestGetAccessTokenLogin(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t, successResp(km, nil))
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	authorized := false
	tok, err := c.GetAccessToken(context.Background(), func() (AuthCodeResult, error) {
		authorized = true
		return codeResult, nil
	})
	if err != nil || tok == "" || !authorized {
		t.Fatalf("login path failed: tok=%q authorized=%v err=%v", tok, authorized, err)
	}
	if (*calls)[0].body.Get("grant_type") != "authorization_code" {
		t.Fatalf("expected authorization_code, got %v", (*calls)[0].body)
	}
}

func TestLaterNonceRetries(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t,
		oauthResp(400, map[string]any{"error": "use_dpop_nonce"}, map[string]string{"DPoP-Nonce": "nonce-1"}),
		successResp(km, map[string]any{"refresh_token": "rt-1"}),
		oauthResp(400, map[string]any{"error": "use_dpop_nonce"}, map[string]string{"DPoP-Nonce": "nonce-2"}),
		successResp(km, nil),
	)
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	set, err := c.ExchangeCode(context.Background(), codeResult)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := c.Refresh(context.Background(), *set); err != nil {
		t.Fatal(err)
	}
	if len(*calls) != 4 {
		t.Fatalf("expected 4 token requests, got %d", len(*calls))
	}
	if proofPayload(t, (*calls)[3].dpop)["nonce"] != "nonce-2" {
		t.Error("second handshake retry should carry nonce-2")
	}
}

func TestTokenDpopHTUOverride(t *testing.T) {
	override := "https://org.okta.com/oauth2/v1/token"
	cfg, km, st := setup(t, override)
	doer, calls := seqDoer(t, successResp(km, nil))
	c := NewTokenClient(cfg, ep, km, st, nil, doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err != nil {
		t.Fatal(err)
	}
	// Still dials the discovered token endpoint...
	if (*calls)[0].url != ep.TokenEndpoint {
		t.Fatalf("dialed %q, want %q", (*calls)[0].url, ep.TokenEndpoint)
	}
	// ...but the proof htu is the override.
	wantHTU, _ := dpop.CanonicalHTU(override)
	if proofPayload(t, (*calls)[0].dpop)["htu"] != wantHTU {
		t.Fatalf("proof htu = %v, want %s", proofPayload(t, (*calls)[0].dpop)["htu"], wantHTU)
	}
}

func TestMissingNonceHeaderWarns(t *testing.T) {
	cfg, km, st := setup(t, "")
	doer, calls := seqDoer(t,
		oauthResp(400, map[string]any{"error": "use_dpop_nonce"}, nil), // no DPoP-Nonce header
		oauthResp(400, map[string]any{"error": "use_dpop_nonce"}, nil),
	)
	buf := &bytes.Buffer{}
	c := NewTokenClient(cfg, ep, km, st, logx.NewWith(buf, "warn"), doer)
	if _, err := c.ExchangeCode(context.Background(), codeResult); err == nil || !strings.Contains(err.Error(), "use_dpop_nonce") {
		t.Fatalf("expected use_dpop_nonce failure, got %v", err)
	}
	if !strings.Contains(buf.String(), "oauth.nonce.missing_header") {
		t.Fatalf("expected missing_header warning; got %q", buf.String())
	}
	if len(*calls) != 2 {
		t.Fatalf("expected 2 requests, got %d", len(*calls))
	}
	if _, ok := proofPayload(t, (*calls)[1].dpop)["nonce"]; ok {
		t.Error("retry proof should have no nonce when the header was stripped")
	}
}
